### DISCUSSION TITLE
How to Deal with the Interfaces Between Different Materials

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#27711] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

In a comment on one of my previous questions, @lindsayad told me (thanks again) that 

> The traditional moose material system has no support for nodal material property evaluation. The on-the-fly functor system does allow nodal evaluations, but it requires the user to specify a subdomain ID. This is to eliminate ambiguity about how to evaluate potentially discontinuous properties at subdomain interfaces.

However, in MOOSE's continuous finite element calculations, each face is executed only once, even though internal faces are shared by two elements. Once a face has been executed in one element, it will not be executed in the neighboring element. If a face is on the interface between different subdomains, wouldn't it only use the material data from one subdomain for its calculations? 
```
template <typename RangeType>
bool
ThreadedElementLoopBase<RangeType>::shouldComputeInternalSide(const Elem & elem,
                                                              const Elem & neighbor) const
{
  auto level = [this](const auto & elem_arg)
  {
    if (_mesh.doingPRefinement())
      return elem_arg.p_level();
    else
      return elem_arg.level();
  };
  const auto elem_id = elem.id(), neighbor_id = neighbor.id();
  const auto elem_level = level(elem), neighbor_level = level(neighbor);

  // When looping over elements and then sides, we need to make sure that we do not duplicate
  // effort, e.g. if a face is shared by element 1 and element 2, then we do not want to do compute
  // work both when we are visiting element 1 *and* then later when visiting element 2. Our rule is
  // to only compute when we are visiting the element that has the lower element id when element and
  // neighbor are of the same adaptivity level, and then if they are not of the same level, then
  // we only compute when we are visiting the finer element
  return (neighbor.active() && (neighbor_level == elem_level) && (elem_id < neighbor_id)) ||
         (neighbor_level < elem_level);
}
```
This seems unreasonable. I'm curious how MOOSE handles this situation. Does anyone have insights? Thank you for your attention.

**Comment by GiudGiud** [comment, marked as answer]
Hello

For faces we actually do have support for material property evaluations. 

You use the getMaterialProperty and getNeighborMaterialProperty

The material property is computed for the current element side quadratures points (index with qp) and same thing for the neighbor element on the other side of the face

Guillaume
